lib/wit/dependency.py

```diff
@@ -68,6 +66,23 @@ def resolve_deps(self, wsroot, repo_paths, download, source_map, packages, queue
 
         return source_map, packages, queue, errors
 
+    def _parallel_clone(self, deps, wsroot, repo_paths, download, errors):
+        def do(dep):
+            try:
+                p = Package(dep.name, repo_paths)
+                p.load(wsroot, download, dep.source, dep.specified_revision)
+            except BadSource as e:
+                errors.append(e)
+
+        threads = list()
+        for dep in deps:
+            t = threading.Thread(target=do, args=(dep,))
```

Comment:
One thing that I'm a little afraid of is that we might trip some rate limiting or DoS detection by spawning many threads that all simultaneously try to connect to GitHub. Would it make sense to cap the number of concurrent threads, as well as introduce some command line argument that forces single-threaded execution?

I think the easiest way to do a cap on concurrent threads is to use the `multiprocessing.dummy` module, which is an implementation of the normal Python `multiprocessing` module except that [it is implemented with threads](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing.dummy), and make use of the [`Pool`](https://docs.python.org/3/library/multiprocessing.html#using-a-pool-of-workers) class. You can basically do something like:

```py
with Pool(10) as pool:
    pool.map(do, deps)
```

Reply:
Github's ratelimiting (for API atleast, couldn't find _anything_ for ssh clone) specifies everything per hour per user, so my guess is that spending 7 minutes downloading 100 repos is the same as spending 1 minute, as it tends to specify 'number of events' rather than 'concurrent events' (and in this diff you don't even get max concurrency).
Happy to put a option in to limit it, I do feel we should leave the default as 'as parallel as you can' unless we can query how much of the ratelimit is remaining for the time period I feel like we're artificially limiting ourselves.

Reply:
Looks like there's no specific limit, but also a caveat of "don't overload our servers": https://github.community/t5/GitHub-API-Development-and/Git-clone-limits-using-git-commands-vs-the-API-what-are-they/td-p/33623

I think I'd request that we still put in a parameter for setting a limit, since I am worried that with no limit we could run into a situation in the future where we're under a deadline and suddenly start running into rate limiting issues, and we can't run CI jobs while frantically trying to ship out a patch release of `wit` that adds in that parameter.

I'm mostly OK with leaving the default as "as parallel as you can", since for a user running this interactively, that should be fine, and even if we run into issues, we can tell the user "just set -j10" or something. The parameter is also important for CI, where the amount of typing/user experience of the wit CLI is less important as it should be captured in a script anyway, so we can afford to have more options.